Runtime Error when Generating

Hi,
really cool tool, i just have problems getting it to work on my Terraform state. It has the Azurerm provider. I am on Windows 10. When trying to inframap generate, i get the following error:
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x20 pc=0x988d24]

goroutine 1 [running]:
github.com/cycloidio/inframap/printer/dot.Dot.Print(0x0, 0x1, 0x1130740, 0xc000006018, 0xc000124620, 0x1029f70)
        /tmp/build/6b18b934/code/printer/dot/printer.go:35 +0x2c4
github.com/cycloidio/inframap/cmd.glob..func1(0x1699ea0, 0xc000056ee0, 0x1, 0x1, 0x0, 0x0)
        /tmp/build/6b18b934/code/cmd/generate.go:84 +0x12c
github.com/spf13/cobra.(*Command).execute(0x1699ea0, 0xc000056d10, 0x1, 0x1, 0x1699ea0, 0xc000056d10)
        /tmp/build/6b18b934/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:852 +0x45a
github.com/spf13/cobra.(*Command).ExecuteC(0x169a3a0, 0x0, 0xa0eb80, 0xc000024178)
        /tmp/build/6b18b934/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x350
github.com/spf13/cobra.(*Command).Execute(...)
        /tmp/build/6b18b934/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/cycloidio/inframap/cmd.Execute(...)
        /tmp/build/6b18b934/code/cmd/root.go:27
main.main()
        /tmp/build/6b18b934/code/main.go:10 +0x35
If you need anything else, let me know.

I have a similar issue:
```
cat *.tf | inframap generate --hcl
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1545e2f]

goroutine 1 [running]:
github.com/cycloidio/inframap/printer/dot.Dot.Print(0x0, 0x1, 0x1d0d920, 0xc000134008, 0x8, 0x0)
        github.com/cycloidio/inframap/printer/dot/printer.go:35 +0x2cf
github.com/cycloidio/inframap/cmd.glob..func1(0x2198760, 0xc00094ee00, 0x0, 0x1, 0x0, 0x0)
        github.com/cycloidio/inframap/cmd/generate.go:84 +0x128
github.com/spf13/cobra.(*Command).execute(0x2198760, 0xc00094edf0, 0x1, 0x1, 0x2198760, 0xc00094edf0)
        github.com/spf13/cobra@v1.1.3/command.go:852 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0x2198c60, 0x0, 0xffffffff, 0xc00010e058)
        github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.1.3/command.go:897
github.com/cycloidio/inframap/cmd.Execute(...)
        github.com/cycloidio/inframap/cmd/root.go:27
main.main()
        github.com/cycloidio/inframap/main.go:10 +0x2e
```

```
cat *.tf | inframap generate --hcl --raw
strict digraph G {
        "aws_route_table_association.private"->"aws_subnet.private";
        "aws_route_table_association.private"->"aws_route_table.private";
        "aws_ecs_task_definition.app"->"aws_iam_role.ecs_task_execution_role";
        "aws_security_group.lb"->"aws_vpc.main";
        "aws_iam_role_policy_attachment.ecs_task_execution_role"->"aws_iam_role.ecs_task_execution_role";
        "aws_appautoscaling_policy.down"->"aws_ecs_cluster.main";
        "aws_appautoscaling_policy.down"->"aws_ecs_service.main";
        "aws_appautoscaling_policy.down"->"aws_appautoscaling_target.target";
        "aws_iam_role_policy_attachment.ecs_dynamodb_policy"->"aws_iam_role.ecs_task_execution_role";
        "aws_iam_role_policy_attachment.ecs_dynamodb_policy"->"aws_iam_policy.dynamodb";
        "aws_internet_gateway.gw"->"aws_vpc.main";
        "aws_subnet.private"->"aws_vpc.main";
        "aws_route.internet_access"->"aws_vpc.main";
        "aws_route.internet_access"->"aws_internet_gateway.gw";
        "aws_security_group.ecs_tasks"->"aws_security_group.lb";
        "aws_security_group.ecs_tasks"->"aws_vpc.main";
        "aws_cloudwatch_metric_alarm.service_cpu_low"->"aws_appautoscaling_policy.down";
        "aws_cloudwatch_metric_alarm.service_cpu_low"->"aws_ecs_cluster.main";
        "aws_cloudwatch_metric_alarm.service_cpu_low"->"aws_ecs_service.main";
        "aws_alb_target_group.app"->"aws_vpc.main";
        "aws_alb_listener.front_end"->"aws_alb.main";
        "aws_alb_listener.front_end"->"aws_alb_target_group.app";
        "aws_route_table.private"->"aws_vpc.main";
        "aws_route_table.private"->"aws_nat_gateway.gw";
        "aws_nat_gateway.gw"->"aws_eip.gw";
        "aws_nat_gateway.gw"->"aws_subnet.public";
        "aws_ecr_lifecycle_policy.main"->"aws_ecr_repository.main";
        "aws_appautoscaling_policy.up"->"aws_ecs_cluster.main";
        "aws_appautoscaling_policy.up"->"aws_ecs_service.main";
        "aws_appautoscaling_policy.up"->"aws_appautoscaling_target.target";
        "aws_subnet.public"->"aws_vpc.main";
        "aws_ecs_service.main"->"aws_alb_target_group.app";
        "aws_ecs_service.main"->"aws_alb_listener.front_end";
        "aws_ecs_service.main"->"aws_iam_role_policy_attachment.ecs_task_execution_role";
        "aws_ecs_service.main"->"aws_ecs_cluster.main";
        "aws_ecs_service.main"->"aws_ecs_task_definition.app";
        "aws_ecs_service.main"->"aws_security_group.ecs_tasks";
        "aws_ecs_service.main"->"aws_subnet.private";
        "aws_cloudwatch_log_stream.log_stream"->"aws_cloudwatch_log_group.log_group";
        "aws_eip.gw"->"aws_internet_gateway.gw";
        "aws_alb.main"->"aws_security_group.lb";
        "aws_alb.main"->"aws_subnet.public";
        "aws_cloudwatch_metric_alarm.service_cpu_high"->"aws_appautoscaling_policy.up";
        "aws_cloudwatch_metric_alarm.service_cpu_high"->"aws_ecs_cluster.main";
        "aws_cloudwatch_metric_alarm.service_cpu_high"->"aws_ecs_service.main";
        "aws_appautoscaling_target.target"->"aws_ecs_cluster.main";
        "aws_appautoscaling_target.target"->"aws_ecs_service.main";
        "aws_alb.main" [ shape=ellipse ];
        "aws_alb_listener.front_end" [ shape=ellipse ];
        "aws_alb_target_group.app" [ shape=ellipse ];
        "aws_appautoscaling_policy.down" [ shape=ellipse ];
        "aws_appautoscaling_policy.up" [ shape=ellipse ];
        "aws_appautoscaling_target.target" [ shape=ellipse ];
        "aws_cloudwatch_log_group.log_group" [ shape=ellipse ];
        "aws_cloudwatch_log_stream.log_stream" [ shape=ellipse ];
        "aws_cloudwatch_metric_alarm.service_cpu_high" [ shape=ellipse ];
        "aws_cloudwatch_metric_alarm.service_cpu_low" [ shape=ellipse ];
        "aws_ecr_lifecycle_policy.main" [ shape=ellipse ];
        "aws_ecr_repository.main" [ shape=ellipse ];
        "aws_ecs_cluster.main" [ shape=ellipse ];
        "aws_ecs_service.main" [ shape=ellipse ];
        "aws_ecs_task_definition.app" [ shape=ellipse ];
        "aws_eip.gw" [ shape=ellipse ];
        "aws_iam_policy.dynamodb" [ shape=ellipse ];
        "aws_iam_role.ecs_task_execution_role" [ shape=ellipse ];
        "aws_iam_role_policy_attachment.ecs_dynamodb_policy" [ shape=ellipse ];
        "aws_iam_role_policy_attachment.ecs_task_execution_role" [ shape=ellipse ];
        "aws_internet_gateway.gw" [ shape=ellipse ];
        "aws_nat_gateway.gw" [ shape=ellipse ];
        "aws_route.internet_access" [ shape=ellipse ];
        "aws_route_table.private" [ shape=ellipse ];
        "aws_route_table_association.private" [ shape=ellipse ];
        "aws_security_group.ecs_tasks" [ shape=rectangle ];
        "aws_security_group.lb" [ shape=rectangle ];
        "aws_subnet.private" [ shape=ellipse ];
        "aws_subnet.public" [ shape=ellipse ];
        "aws_vpc.main" [ shape=ellipse ];

}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,11 @@
 - Terraform version from 0.14.9 to 0.14.10
   ([PR #128](https://github.com/cycloidio/inframap/pull/128))
 
+### Fixed
+
+- When generating code from HCL errors are now displayed properly
+  ([Issue #120](https://github.com/cycloidio/inframap/pull/120))
+
 ## [0.6.2] _2021-03-30_
 
 ### Added

--- a/cmd/generate.go
+++ b/cmd/generate.go
@@ -50,7 +50,8 @@ var (
 					fs := afero.NewMemMapFs()
 					path = "module.tf"
 
-					f, err := fs.Create(path)
+					var f afero.File
+					f, err = fs.Create(path)
 					if err != nil {
 						return err
 					}

--- a/errcode/code.go
+++ b/errcode/code.go
@@ -23,8 +23,9 @@ var (
 	ErrProviderNotFoundDataSource = errors.New("provider data source not found")
 	ErrProviderNotFound           = errors.New("provider not found")
 
-	ErrInvalidTFStateFile    = errors.New("invalid Terraform State file")
-	ErrInvalidTFStateVersion = errors.New("invalid Terraform State version, we only support version 3 and 4")
+	ErrInvalidTFStateFile                  = errors.New("invalid Terraform State file")
+	ErrInvalidTFStateVersion               = errors.New("invalid Terraform State version, we only support version 3 and 4")
+	ErrInvalidTFStateFileMissingResourceID = errors.New("invalid Terraform State file, a resource is missing the attributes.id")
 
 	ErrPrinterNotFound = errors.New("printer not found")
 

--- a/generate/state.go
+++ b/generate/state.go
@@ -143,10 +143,14 @@ func FromState(tfstate json.RawMessage, opt Options) (*graph.Graph, map[string]i
 				if err != nil {
 					return nil, nil, err
 				}
+				tfid, ok := aux["id"]
+				if !ok {
+					return nil, nil, fmt.Errorf("resource %q: %w", rk, errcode.ErrInvalidTFStateFileMissingResourceID)
+				}
 				n := &graph.Node{
 					ID:        uuid.NewV4().String(),
 					Canonical: prefixWithModule(moduleMode, m.Addr.Module().String(), rk),
-					TFID:      aux["id"].(string),
+					TFID:      tfid.(string),
 					Resource:  *res,
 				}
 

--- a/generate/state.go
+++ b/generate/state.go
@@ -239,27 +239,29 @@ func migrateVersions(src []byte, f *statefile.File) error {
 			return fmt.Errorf("error while reading TFState: %w", err)
 		}
 
-		for _, m := range state["modules"].([]interface{}) {
-			rs, ok := m.(map[string]interface{})["resources"]
-			if !ok {
-				continue
-			}
-			for rn, rv := range rs.(map[string]interface{}) {
-				// As we have replaced from 'depends_on' to 'dependencies'
-				// we have to access it with 'dependencies'
-				dps, ok := rv.(map[string]interface{})["dependencies"]
-				if !ok && dps == nil {
+		if _, ok := state["modules"]; ok {
+			for _, m := range state["modules"].([]interface{}) {
+				rs, ok := m.(map[string]interface{})["resources"]
+				if !ok {
 					continue
 				}
-				for _, d := range dps.([]interface{}) {
-					ds := strings.Split(d.(string), ".")
-					deps[rn] = append(deps[rn], addrs.ConfigResource{
-						Resource: addrs.Resource{
-							Mode: addrs.ManagedResourceMode,
-							Type: ds[0],
-							Name: ds[1],
-						},
-					})
+				for rn, rv := range rs.(map[string]interface{}) {
+					// As we have replaced from 'depends_on' to 'dependencies'
+					// we have to access it with 'dependencies'
+					dps, ok := rv.(map[string]interface{})["dependencies"]
+					if !ok && dps == nil {
+						continue
+					}
+					for _, d := range dps.([]interface{}) {
+						ds := strings.Split(d.(string), ".")
+						deps[rn] = append(deps[rn], addrs.ConfigResource{
+							Resource: addrs.Resource{
+								Mode: addrs.ManagedResourceMode,
+								Type: ds[0],
+								Name: ds[1],
+							},
+						})
+					}
 				}
 			}
 		}

--- a/generate/state_test.go
+++ b/generate/state_test.go
@@ -2,6 +2,7 @@ package generate_test
 
 import (
 	"errors"
+	"fmt"
 	"io/ioutil"
 	"testing"
 
@@ -28,6 +29,14 @@ func TestFromState(t *testing.T) {
 		require.NoError(t, err)
 		require.NotNil(t, g)
 		require.NotNil(t, cfg)
+	})
+	t.Run("NoID", func(t *testing.T) {
+		src, err := ioutil.ReadFile("./testdata/error_missing_id.json")
+		require.NoError(t, err)
+
+		_, _, err = generate.FromState(src, generate.Options{Clean: true, Connections: true, ExternalNodes: true})
+		fmt.Println(err)
+		assert.True(t, errors.Is(err, errcode.ErrInvalidTFStateFileMissingResourceID))
 	})
 }
 

--- a/generate/state_test.go
+++ b/generate/state_test.go
@@ -20,6 +20,15 @@ func TestFromState(t *testing.T) {
 		_, _, err = generate.FromState(src, generate.Options{Clean: true, Connections: true, ExternalNodes: true})
 		assert.True(t, errors.Is(err, errcode.ErrInvalidTFStateVersion))
 	})
+	t.Run("Empty", func(t *testing.T) {
+		src, err := ioutil.ReadFile("./testdata/empty.json")
+		require.NoError(t, err)
+
+		g, cfg, err := generate.FromState(src, generate.Options{Clean: true, Connections: true, ExternalNodes: true})
+		require.NoError(t, err)
+		require.NotNil(t, g)
+		require.NotNil(t, cfg)
+	})
 }
 
 func TestFromState_AWS(t *testing.T) {

--- a/generate/testdata/empty.json
+++ b/generate/testdata/empty.json
@@ -1,0 +1,7 @@
+{
+    "version": 3,
+    "terraform_version": "0.11.14",
+    "serial": 83,
+    "lineage": "3fd5eb87-d6b9-8b47-6c2d-7209f7f884ee",
+    "resources": []
+}

--- a/generate/testdata/error_missing_id.json
+++ b/generate/testdata/error_missing_id.json
@@ -1,0 +1,18 @@
+{
+  "version":4,
+  "terraform_verison": "0.12.28",
+  "resources": [
+    {
+      "type":"aws_db_instance",
+      "mode":"managed",
+      "name": "mydb",
+      "provider": "provider.aws",
+      "instances": [
+        {
+          "attributes":{}
+        }
+      ]
+    }
+  ]
+}
+

